tests/themis/wrappers/android/com/cossacklabs/themis/test/SymmetricKeyTest.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (c) 2019 Cossack Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.cossacklabs.themis.test;
+
+import com.cossacklabs.themis.SymmetricKey;
+import com.cossacklabs.themis.InvalidArgumentException;
+import com.cossacklabs.themis.NullArgumentException;
+
+import static org.junit.Assert.*;
+import org.junit.Test;
+
+public class SymmetricKeyTest {
+    @Test
+    public void generateNewKey() {
+        SymmetricKey key = new SymmetricKey();
+
+        byte[] keyBytes = key.toByteArray();
+        assertNotNull(keyBytes);
+        assertTrue(keyBytes.length > 0);
+    }
+
+    @Test
+    public void restoreKeyFromBytes() {
+        byte[] buffer = { 3, 14, 15, 92, 6 };
+        SymmetricKey key = new SymmetricKey(buffer);
```

Comment:
do we really want to allow users to create their own 'insecure' keys?

Reply:
In the world where we provide strict type interfaces (accepting a concrete `SymmetricKey` or `PublicKey` instead of just `byte[]`) the users will need a way to construct keys from raw bytes that they read from a file, for example. The constructors should ensure that the keys are only constructed from valid bytes.

In the future we may impose additional restrictions to prohibit usage of weak keys. But right now we don’t have any other than keys being non-empty.

Plus, `PublicKey` and `PrivateKey` in Java already allow construction from byte arrays. Though, they don’t validate them at all, leaving that task to crypto systems.

Reply:
My point is that if we will have two interfaces (_read this as illustration_):

- `encryptUsingKey(SymmetricKey: key, ...)`
- `encryptUsingPassphrase(String: passphrase, ...)`

users might do following:

```
key = SymmetricKey("qwerty".toBytesArray())
cell.encryptUsingKey(key, data)
```

thus using low entropy string as input param for creating bad "symmetric key", avoiding KDF call.

I'd suggest separate two interfaces:
- `SymmetricKey` could be only generated by Soter, returns 32 bytes
- `Passphrase` could be generated from user input

Reply:
They sure can do that, but there is no realistic way to prevent that situation with just API design. We cannot gauge 'randomness' of a key by a single sample. We could introduce a bottom limit on the key length, but that will not prevent the users from using `"could you please let me use Secure Cell, mkay?".toByteArray()` as a key which is hardly random. Even _if_ it's random, that's a hardcoded, fixed key that can be reverse-engineered or just stolen from the source code.

You're right about the passphrase interface, but we cannot leave the users with only a constructor that returns new instances of `SymmetricKey`. This would mean that the key can never leave local memory of the JVM instance that created it. The users will not be able to save the key to a persistent storage, then retrieve it later and use to decrypt the data. For that they _will_ need to way to convert arbitrary `byte[]` array into a key usable by Secure Cell. That is, without a constructor like that.